lib/pid/src/PID.cpp

```diff
@@ -1,492 +1,120 @@
-  } else {
-    tempTauR = (1.0 / tauI) * tSample_.count();
-  }
-
-  // For "bumpless transfer" we need to rescale the accumulated error.
-  if (inAuto) {
-    if (tempTauR == 0.0) {
-      accError_ = 0.0;
-    } else {
-      accError_ *= (Kc_ * tauR_) / (Kc * tempTauR);
-    }
-  }
-
-  Kc_   = Kc;
-  tauR_ = tempTauR;
-  tauD_ = tauD / tSample_.count();
+void PID::PID::updateDeadzone(float deadzone) {
+  std::lock_guard<Mutex> lock(m_mutex);
+  m_deadzone = deadzone;
 }
 
-void PID::reset(void) {
-  float scaledBias = 0.0;
-
-  if (usingFeedForward) {
-    scaledBias = (bias_ - outMin_) / outSpan_;
-  } else {
-    scaledBias = (realOutput_ - outMin_) / outSpan_;
-  }
-
-  prevControllerOutput_ = scaledBias;
-  prevProcessVariable_  = (processVariable_ - inMin_) / inSpan_;
-
-  // Clear any error in the integral.
-  accError_ = 0;
-}
-
-void PID::setMode(int mode) {
-  // We were in manual, and we just got set to auto.
-  // Reset the controller internals.
-  if (mode != 0 && !inAuto) {
-    reset();
-  }
-
-  inAuto = (mode != 0);
+float PID::PID::reportProportionalGain() const {
+  std::lock_guard<Mutex> lock(m_mutex);
+  return m_PGain;
 }
 
-void PID::setInterval(std::chrono::duration<float> interval) {
-  if (interval > 0s) {
-    // Convert the time-based tunings to reflect this change.
-    tauR_ *= (interval / tSample_);
-    accError_ *= (tSample_ / interval);
-    tauD_ *= (interval / tSample_);
-    tSample_ = interval;
-  }
-}
-
-void PID::setSetPoint(float sp) {
-  setPoint_ = sp;
+float PID::PID::reportIntegralGain() const {
+  std::lock_guard<Mutex> lock(m_mutex);
+  return m_IGain;
 }
 
-void PID::setProcessValue(float pv) {
-  processVariable_ = pv;
+float PID::PID::reportDerivativeGain() const {
+  std::lock_guard<Mutex> lock(m_mutex);
+  return m_DGain;
 }
 
-void PID::setBias(float bias) {
-  bias_            = bias;
-  usingFeedForward = 1;
+float PID::PID::reportDeadzone() const {
+  std::lock_guard<Mutex> lock(m_mutex);
+  return m_deadzone;
 }
 
-void PID::setDeadZoneError(float error) {
-  deadZoneError_ = error;
+void PID::PID::reset() {
+  std::lock_guard<Mutex> lock(m_mutex);
+  m_IAccumulator = 0;
+  m_pastError    = 0;
+  m_pastPV       = 0;
+  m_timer.stop();
+  m_timer.reset();
 }
 
-void PID::setRealOutput(float realOutput) {
-  realOutput_ = realOutput;
+float PID::PID::computePPath(float error) {
+  // no mutex lock needed since inside compute() only
+  return error * m_PGain;
 }
 
-void PID::setupAutoTune(void *outputPointer, float *inputPointer, int actuatorType) {
-  // set actuator type
-  actuatorType_ = actuatorType;
-
-  // set I/O pointers
-  output_ = outputPointer;
-  input_  = inputPointer;
+float PID::PID::computeIPath(float error, int64_t dt) {
+  // no mutex lock needed since inside compute() only
+  m_IAccumulator += error * dt * m_IGain;
+  m_IAccumulator = std::clamp(m_IAccumulator, static_cast<float>(m_lowerBound), static_cast<float>(m_upperBound));
+  return m_IAccumulator;
```

Comment:
still find it weird that you're returning a member variable here. other member functions already have access to member variables which is why if you return a variable here, i suggest you return the amount the ipath has changed.